From 28c579dc30a7e5499d68a128ec479f9c064c1dd3 Mon Sep 17 00:00:00 2001
From: MagneFire <IDaNLContact@gmail.com>
Date: Thu, 16 Jul 2020 22:51:52 +0200
Subject: [PATCH] Add wakeup event for use with ambient mode. It seems like
 there is no easy way to get a callback from when an event occurs. The only
 way is using voland. Which, on AsteroidOS, is used by the alarmclock app. So
 scheduling an event for use with ambient mode would actually show the
 alarmclock dialog (and update the screen). This is due to alarmclock using a
 D-Bus (systemd) service that triggers on any voland signal. I don't think
 there is a way to filter these events with systemd. Another thing that may be
 used is the `next_bootup_event` event. But this is not a reliable source,
 because alarmclock events also cause this signal, can cause multiple triggers
 at irregular intervals.

Somehow we need to update the screen without triggering the alarmclock alarm dialog.
The most straightforward way to do this is by adding a special `wakeup` type event, that doesn't trigger the voland event, but instead a new event.
---
 src/server/adaptor.h   |  1 +
 src/server/machine.cpp |  1 +
 src/server/machine.h   |  1 +
 src/server/state.cpp   | 17 +++++++++++++++--
 src/server/state.h     |  1 +
 src/server/timed.cpp   |  1 +
 src/server/timed.h     |  1 +
 7 files changed, 21 insertions(+), 2 deletions(-)

diff --git a/src/server/adaptor.h b/src/server/adaptor.h
index 7c28921..f7a6491 100644
--- a/src/server/adaptor.h
+++ b/src/server/adaptor.h
@@ -80,6 +80,7 @@ signals:
   void settings_changed_1(bool) ;
   void next_bootup_event(int next_boot_event, int next_non_boot_event);
   void alarm_triggers_changed(Maemo::Timed::Event::Triggers);
+  void wakeup_event();

 public slots:

diff --git a/src/server/machine.cpp b/src/server/machine.cpp
index 6977fb0..dee03cf 100644
--- a/src/server/machine.cpp
+++ b/src/server/machine.cpp
@@ -116,6 +116,7 @@ machine_t::machine_t(const Timed *daemon) : timed(daemon)
   state_armed->open() ;

   QObject::connect(state_dlg_wait, SIGNAL(voland_needed()), this, SIGNAL(voland_needed())) ;
+  QObject::connect(state_dlg_wait, SIGNAL(wakeup_event()), this, SIGNAL(wakeup_event())) ;

   QObject::connect(state_dlg_wait, SIGNAL(closed()), state_dlg_requ, SLOT(open())) ;
   QObject::connect(state_dlg_wait, SIGNAL(closed()), state_dlg_user, SLOT(open())) ;
diff --git a/src/server/machine.h b/src/server/machine.h
index de0098f..314aba7 100644
--- a/src/server/machine.h
+++ b/src/server/machine.h
@@ -101,6 +101,7 @@ Q_SIGNALS:
   void queue_to_be_saved() ;
   void voland_needed() ;
   void next_bootup_event(int, int);
+  void wakeup_event();
   void child_created(unsigned, int) ;
   void alarm_present(bool present);
   void alarm_trigger(QMap<QString, QVariant> triggers);
diff --git a/src/server/state.cpp b/src/server/state.cpp
index 02a7d74..f840a74 100644
--- a/src/server/state.cpp
+++ b/src/server/state.cpp
@@ -738,8 +738,21 @@ void state_aborted_t::enter(event_t *e)
 void state_dlg_wait_t::enter(event_t *e)
 {
   e->flags |= EventFlags::In_Dialog ;
-  if (not is_open)
-    emit voland_needed() ;
+  for(attribute_t::const_iterator at=e->attr.txt.begin(); at!=e->attr.txt.end(); at++)
+  {
+    QString key = string_std_to_q(at->first) ;
+    QString val = string_std_to_q(at->second) ;
+    if (key == "type") {
+      if (val == "wakeup") {
+        machine->state_served->go_to(e);
+        emit wakeup_event();
+      } else {
+        if (not is_open)
+          emit voland_needed();
+      }
+      break;
+    }
+  }
   abstract_gate_state_t::enter(e) ;
 }

diff --git a/src/server/state.h b/src/server/state.h
index 92046e5..0e84ccb 100644
--- a/src/server/state.h
+++ b/src/server/state.h
@@ -314,6 +314,7 @@ public:
   uint32_t cluster_bits() { return EventFlags::Cluster_Dialog ; }
 Q_SIGNALS:
   void voland_needed() ;
+  void wakeup_event();
 private:
   Q_OBJECT ;
 } ;
diff --git a/src/server/timed.cpp b/src/server/timed.cpp
index 1d1179a..aa386d0 100644
--- a/src/server/timed.cpp
+++ b/src/server/timed.cpp
@@ -316,6 +316,7 @@ void Timed::init_create_event_machine()

   // Forward signal from am to DBUS via com_nokia_time DBUS adaptor
   QObject::connect(am, SIGNAL(next_bootup_event(int,int)), this, SIGNAL(next_bootup_event(int,int)));
+  QObject::connect(am, SIGNAL(wakeup_event()), this, SIGNAL(wakeup_event()));
   voland_watcher = NULL ;
   QObject::connect(this, SIGNAL(voland_registered()), am, SIGNAL(voland_registered())) ;
   QObject::connect(this, SIGNAL(voland_unregistered()), am, SIGNAL(voland_unregistered())) ;
diff --git a/src/server/timed.h b/src/server/timed.h
index e7cb2a7..2bdd62b 100644
--- a/src/server/timed.h
+++ b/src/server/timed.h
@@ -165,6 +165,7 @@ Q_SIGNALS:
   void settings_changed(const Maemo::Timed::WallClock::Info &, bool system_time) ;
   void next_bootup_event(int next_boot_event, int next_non_boot_event);
   void alarm_triggers_changed(Maemo::Timed::Event::Triggers);
+  void wakeup_event();
   // void settings_changed_1(bool system_time) ;
 public:
   Timed(int ac, char **av) ;
--
2.28.0
